### PR TITLE
[DTensor] initial support for decomps + sharding prop

### PR DESCRIPTION
**Pull request by pianpwk**
Following @tianyu-l's https://github.com/pytorch/pytorch/pull/130887

Adds support for ops with no sharding prop strategy, but a registered decomposition.

Now if sharding prop sees a decomposable op, it:
1. Runs the decomposed op under a custom TorchDispatchMode, which propagates the placements as side information (initially used a make_fx implementation, but this required a threading lock as it relies on [global state](https://github.com/pytorch/pytorch/blob/2a26c9a32661ee2b4b049e3bd1b889fc3af30880/torch/fx/_symbolic_trace.py#L1167))
2. Enumerates potential input placement combinations based on the actual input placements, on a single-dim mesh, then for each of them, propagates through torch_dispatch via sharding prop, while banning any intermediate redistributions.
3. Returns the expanded full-mesh strategy from the filtered strategies.

Some caveats:
- Since the dispatch mode runs sharding prop, the shard prop cache should kick in, both in the normal case (running the same op twice), and also when we recursively decompose (if op1 -> op2 -> some decomp, running op1 caches for op2).
- One common failure case is decompositions calling factory methods (e.g. [torch.ones, torch.arange](https://github.com/pytorch/pytorch/blob/41f42a0fc3ea1fbfdf05b4c030d7df815bdfe19d/torch/_decomp/decompositions.py#L818-L821)). The main problem seems to be assigning placements to these tensors, and it's not so obvious what their placements should be, especially when they might take in sharded sizes, and we can't completely detect when this is the case. For now, intermediate shard prop will fail (no sharding strategy; they don't take DTensor inputs), but a potential future improvement is to permit the full-Replicate case for these graphs.
- Sharding prop is currently via a `propagate_op_sharding` call, on explicit placement types. Once [single-dim strategy](https://github.com/pytorch/pytorch/pull/167677) coverage is broader, this should be doable on _ShardPlaceholders instead, making the enumeration & propagation process cheaper, though maybe more manual.
- (Maybe hackily) uses a fake 1-rank 1d mesh to do single-dim propagation

Removes the following xfails (+some more aten ops with decomp coverage, but still failing tests):
```
__rsub__
addmv
addr
alias_copy
all
any
count_nonzero
dist
expand_copy
fill
floor_divide
index_select
linalg.vecdot
masked_fill
mv
nn.functional.celu
nn.functional.channel_shuffle
nn.functional.elu
nn.functional.hardsigmoid
nn.functional.hardswish
nn.functional.hardtanh
nn.functional.leaky_relu
nn.functional.logsigmoid
nn.functional.margin_ranking_loss
nn.functional.mish
nn.functional.multilabel_soft_margin_loss
nn.functional.pairwise_distance
nn.functional.pixel_shuffle
nn.functional.pixel_unshuffle
nn.functional.prelu
nn.functional.relu6
nn.functional.selu
nn.functional.softplus
nn.functional.softshrink
nn.functional.triplet_margin_loss
nn.functional.triplet_margin_with_distance_loss
permute_copy
rsub
t_copy
trace
vdot
view_copy
```

### CODE REVIEW

File: torch/distributed/tensor/_decompositions.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+        single_dim_strategies = []
+        interp = DecompShardingInterpreter(graph, mesh.device_type, sharding_prop)
+        for placement in placements:
+            try:
+                output = cls._propagate_through_decomp(interp, placement, op_schema)
+            except NotImplementedError:
+                # immediately return; some op doesn't have a sharding strategy.
+                return None
+            except (ImplicitRedistributionError, RuntimeError):
+                # I feel like we shouldn't have to catch RuntimeErrors?
```
Comment (wconstab) [inline torch/distributed/tensor/_decompositions.py]: View ops will raise if they can't be performed locally - for example if you take a sharded tensor and reshape it, its possible the resulting global tensor shape does not match with the local data and the placement types we have defined.  We could make this a more specific type of error. cc @weifengpy

File: test/distributed/tensor/test_decompositions.py [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+        self.assertEqual(out.placements, (Partial("avg"), Partial("avg")))
+
+        # expand_copy
+        check_no_strategy(aten.expand_copy.default)
+        input = d_empty(16, 1, device_mesh=mesh, placements=[Partial("min")])
+        out = aten.expand_copy.default(input, [-1, 16])
+        self.assertEqual(out.placements, (Partial("min"),))
+
+        # # index_add
+        # # index_put decomp errors; https://github.com/pytorch/pytorch/issues/170934
```
Comment (wconstab) [inline test/distributed/tensor/test_decompositions.py]: any reason this is commented?

File: test/distributed/tensor/test_decompositions.py [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+        self.assertEqual(out.placements, (Replicate(),))
+
+        # polar: force replicate
+        check_no_strategy(aten.polar.default)
+        x = d_empty(16, device_mesh=mesh, placements=[Partial()])
+        y = d_empty(16, device_mesh=mesh, placements=[Partial()])
+        out = aten.polar.default(x, y)
+        self.assertEqual(out.placements, (Replicate(),))
+
+        # # pixel_shuffle
```
Comment (wconstab) [inline test/distributed/tensor/test_decompositions.py]: same

**Comment by wconstab** [comment]
this PR probably should have removed a handful of xfails from test_dtensor_ops.py - did you just not look into that yet?

File: test/distributed/tensor/test_decompositions.py (hunk 2 above)
Comment (pianpwk) [reply test/distributed/tensor/test_decompositions.py]: the underlying strategy is incorrect... I might just delete this test

File: test/distributed/tensor/test_decompositions.py (hunk 3 above)
Comment (pianpwk) [reply test/distributed/tensor/test_decompositions.py]: the replicate behavior was unexpected (may also remove this)

File: torch/distributed/tensor/_decompositions.py [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+        sharding_prop: ShardingPropagator,
+    ):
+        super().__init__(module)
+        self.sharding_prop = sharding_prop
+
+        # Build single mesh dim strategies using fake 1d mesh
+        import torch.distributed as dist
+
+        world_size = dist.get_world_size() if dist.is_initialized() else 2
+        self.mesh = DeviceMesh(mesh_device, torch.arange(world_size))
```
Comment (pianpwk) [inline torch/distributed/tensor/_decompositions.py]: @wconstab this part is still suspect to me, I had to do this to avoid the assertion in https://github.com/pytorch/pytorch/blob/c22397c4addd9acc4bb6414b6b0510e3ba090d1c/torch/distributed/tensor/_redistribute.py#L302

do you have a concept of what the device mesh should be, if we're propagating through the decomp graph with single-dim strategies?

Comment (wconstab) [reply torch/distributed/tensor/_decompositions.py]: wondering if you hit the same issue as i ran into here: https://github.com/pytorch/pytorch/pull/172478
still ruminating on what the best solution is

Comment (wconstab) [reply torch/distributed/tensor/_decompositions.py]: to answer your question more literally though,
> do you have a concept of what the device mesh should be, if we're propagating through the decomp graph with single-dim strategies?

I don't think we should need to know about the device mesh when we are evaluating the single-dim portion of the single-dim rule.  However when we get into expansion, we do look at the mesh, and it should be the real mesh, not some single-dim fake mesh.

Comment (pianpwk) [reply torch/distributed/tensor/_decompositions.py]: Yep, exactly that error.

Expansion is definitely on the real mesh, I'm mainly worried that this fake mesh ends up filtering out valid strategies (because say, maybe the mesh dim size is too large)

Comment (wconstab) [reply torch/distributed/tensor/_decompositions.py]: I'm not clear about what fake mesh exists and why its size is off. In the PR I linked there was a unit test doing something with a submesh where not all ranks participate, which in itself is a bit unusual. For normal cases for decomp tracing shouldn't we just be using the runtime mesh sizes?

Comment (pianpwk) [reply torch/distributed/tensor/_decompositions.py]: Ah sorry, patching https://github.com/pytorch/pytorch/pull/172478 actually helps a lot.

I guess because DecompShardingInterpreter is actually calling sharding prop, there needs to be a mesh attached, and at the same time, we're trying to enumerate on single-dim. There's a couple options that seem to work for me:

1) reuse the real n-d mesh, but only search over placements on one of the mesh dims, setting the others as Replicate.
2) create a fake 1d mesh with real world size
3) create a fake 1d mesh with size 2 (or maybe even 1?), and patch redistribute cost to avoid failing on non-participating ranks

I was worried 1) & 2) could incorrectly filter out strategies that would be valid on the real n-d mesh (for 1 we inadvertently choose a mesh dim bad for even sharding (e.g. 5); for 2 maybe the world size is larger than tensor size, and we don't shard). Option 3) seemed like a way to avoid false negatives here.

Comment (pianpwk) [reply torch/distributed/tensor/_decompositions.py]: Ok, I think we've settled on ~3: create a fake 1d mesh with size 1, but avoid using the real PG, and have all processes act like rank 0: `fake_mesh = DeviceMesh(mesh.device_type, [0], _init_backend=False, _rank=0)`

Previously the 0-cost for non-participating ranks was working, but returning cost = 0 for all strategies on ranks >= 1.  I'm not entirely sure why this was "working" - maybe the redistribute planner was only being invoked when we definitely needed redistribution, and while some of those discovered strategies were included with incorrect 0-cost, they weren't chosen anyways because we banned implicit redistributes.

File: torch/distributed/tensor/_decompositions.py [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+class DecompShardingStrategy:
+    """
+    Generates sharding strategies for ops by tracing through their decompositions.
+
+    For each candidate input placement combination, runs the decomposition on meta
+    tensors under PlacementTrackingMode to determine the output placement. These
+    single-dimension strategies are then expanded to the full mesh.
+    """
+
+    @classmethod
```
Comment (zpcore) [inline torch/distributed/tensor/_decompositions.py]: Are reason we don't use staticmethod directly?

File: torch/distributed/tensor/_decompositions.py [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+        # pyrefly: ignore [missing-attribute]
+        self._record_output_specs(out, output_sharding.output_spec)
+        return out
+
+    def _record_output_specs(
+        self, output: Any, output_spec: DTensorSpec | tuple
+    ) -> None:
+        if isinstance(output_spec, DTensorSpec):
+            output._spec = output_spec
+        elif isinstance(output_spec, tuple) and isinstance(output, (tuple, list)):
```
Comment (zpcore) [inline torch/distributed/tensor/_decompositions.py]: Is it possible for output to be nested tuple?

File: torch/distributed/tensor/_decompositions.py [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+            flat = [p for p in result if p is not None]
+            return flat[0] if len(flat) == 1 else tuple(flat)
+        return result
+
+    @classmethod
+    def _get_candidate_placements(
+        cls, op_schema: OpSchema
+    ) -> list[tuple[Placement | None]]:
+        tensor_specs = _extract_input_specs(op_schema)
+        flat_specs, _ = tree_flatten(list(tensor_specs))
```
Comment (zpcore) [inline torch/distributed/tensor/_decompositions.py:217]: Just checking... There are many places trying to flatten op_schema to extract out DTensorSpec. Is it possible to reuse the OpSchema property like https://github.com/pytorch/pytorch/blob/5b897bd3d1e65d0095b44ea40dfad40cb83d60bf/torch/distributed/tensor/_op_schema.py#L374?

File: torch/distributed/tensor/_decompositions.py [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+            op_schema.args_schema + tuple(op_schema.kwargs_schema.values()),
+        )
+
+        # Create a fake mesh where all ranks pretend to be rank 0.
+        # This ensures identical cost computation across all ranks during
+        # decomposition tracing, avoiding potential SPMD divergence.
+        fake_mesh = DeviceMesh(mesh.device_type, [0], _init_backend=False, _rank=0)
+        single_dim_strategies = []
+        output_placements: list[Placement | tuple[Placement, ...]] = []
+        for placement in placements:
```
Comment (zpcore) [inline torch/distributed/tensor/_decompositions.py]: The naming is kind of confusing. The `placement` is actually of type tuple[Placement].

File: torch/distributed/tensor/_decompositions.py [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+
+            options: set[Placement] = {Replicate()}
+            for p in spec.placements:
+                if isinstance(p, (Shard, _StridedShard)):
+                    for i in range(spec.ndim):
+                        options.add(
+                            Shard(i)
+                            if isinstance(p, Shard)
+                            else _StridedShard(i, split_factor=p.split_factor)
+                        )
```
Comment (zpcore) [inline torch/distributed/tensor/_decompositions.py]: Sorry I didn't quite understand this part. It looks like we are trying to find possible placement based on this arg's spec over all mesh dims. Why not just `set(spec.placements+Replicate)`? cc @wconstab

Comment (pianpwk) [reply torch/distributed/tensor/_decompositions.py]: Yeah, I wasn't sure how to handle this. One extreme would be enumerating all R/S/SS/P placements unconditionally, regardless of input placements. `set(spec.placements+Replicate)` is on the other end.

I was partly following https://github.com/pytorch/pytorch/pull/130887, maybe there's a scenario where the input sharding isn't able to produce a no-redistribute rule from decomposition, but sharding on another tensor dim can, and the ideal strategy is alltoall + that rule.

But maybe that's unrealistic?

File: test/distributed/tensor/test_utils.py [hunk 10]
```diff
@@ -1677,7 +1678,9 @@ def test_explicit_matmul(self):
 
             # explicit redistribute works too
             with ExplicitRedistributionContext():
-                with self.assertRaisesRegex(RuntimeError, "Implicit redistribution"):
+                with self.assertRaisesRegex(
+                    ImplicitRedistributionError, "Implicit redistribution"
```
Comment (wconstab) [inline test/distributed/tensor/test_utils.py]: suggest moving this refactor to a separate PR and check it can land safely. (possible someone internal already hardcoded test values on RuntimeError, we should update those if so)

File: torch/distributed/tensor/_decompositions.py (hunk 6 above)
Comment (wconstab) [reply torch/distributed/tensor/_decompositions.py]: regardless, probably better to use tree_map here?

File: torch/distributed/tensor/_decompositions.py (hunk 9 above)
Comment (wconstab) [reply torch/distributed/tensor/_decompositions.py]: hmm.  several thoughts

1) it would be nice to reuse 'shardingplaceholder' here, and have it get filled in later.  that 'filling in' takes care of the problem of figuring out which stridedshard combinations to use.  but since this decomp infra isn't specific to single-dim rules, we can't do that.
2) looking at the current input placements to decide the sharding rule is very 'old sharding rule' style.  the single dim mantra says "you should list all the valid rules, doesn't matter what your input is".  

so, generally i'd prefer if we just enumerate the valid placements rather than looking at inputs, but its impossible to do this for stridedshard.  If we plan to eventually have a specialized decomp infra for single-dim rules that uses shardplaceholder to avoid this issue, i'd be ok with doing the input-influenced rules for now.

But nit: is it cleaner if you just use tree_map to cover the placements and then jam them into a set, without all the if/elif?

File: torch/distributed/tensor/_decompositions.py [hunk 11]
```diff
@@ -0,0 +1,231 @@
...
+
+        candidate_placements = DecompShardingStrategy._get_candidate_placements(
+            op_schema
+        )
+        mesh = try_find_mesh_from_args(
+            op_schema.op,
+            op_schema.args_schema + tuple(op_schema.kwargs_schema.values()),
+        )
+
+        # Create a fake mesh where all ranks pretend to be rank 0.
```
Comment (wconstab) [inline torch/distributed/tensor/_decompositions.py]: i think we better document this a bit better, bc i almost forgot our discussion points

1) the tracing rank0 part makes sense, since all ranks are required to have the same strategies and same outcome from shard prop.  (but to this point, using actual rank shouldn't be any worse should it?)
2) using a fake mesh can lead to false negatives/positives, since sometimes we rule out a sharding due to unevenness on a mesh dim.  We steer towards false positives by using a fake mesh of size 1 (and this should have zero false negatives?).  
- And what do we do about the false positives later?

File: torch/distributed/tensor/_decompositions.py (hunk 7 above)
Comment (pianpwk) [reply torch/distributed/tensor/_decompositions.py:217]: ah, I think this function filters out Nones for non-DTensor args and loses positions

File: torch/distributed/tensor/_decompositions.py (hunk 9 above)
Comment (pianpwk) [reply torch/distributed/tensor/_decompositions.py]: simplified/generalized this a bit, but I guess we still need some special-casing to expand sharding to all tensor dimensions

File: test/distributed/tensor/test_utils.py (hunk 10 above)
Comment (pianpwk) [reply test/distributed/tensor/test_utils.py]: switching this back to RuntimeError for now, it's more of BE

File: torch/distributed/tensor/_decompositions.py (hunk 11 above)
Comment (pianpwk) [reply torch/distributed/tensor/_decompositions.py]: added something

File: torch/distributed/tensor/_decompositions.py [hunk 12]
```diff
@@ -117,9 +110,14 @@ def propagate_strategy(
...
         )
 
-        # Create a fake mesh where all ranks pretend to be rank 0.
-        # This ensures identical cost computation across all ranks during
+        # Create a fake size-1 mesh where all ranks pretend to be rank 0.
+        # This ensures identical strategy computation across all ranks during
         # decomposition tracing, avoiding potential SPMD divergence.
+        #
+        # Using a fake mesh could potentially cause false negatives/positives
+        # (in terms of valid sharding strategies). The size-1 mesh attempts to avoid
```
Comment (wconstab) [inline torch/distributed/tensor/_decompositions.py]: to strengthen this a bit, any reason the fake size-1 mesh doesn't perfectly solve the false negatives?  I can't think of a case where having size >1 expands the number of valid rules, but I haven't thought of a proof for this either.

**Review by wconstab — APPROVED** [review]
lgtm. let's start trying it out more. we should run my sharding validator on the newly supported ops and see if it uncovers anything.

File: torch/distributed/tensor/_decompositions.py (hunk 7 above)
Comment (zpcore) [reply torch/distributed/tensor/_decompositions.py:217]: I see, good to know!

File: torch/distributed/tensor/_decompositions.py (hunk 9 above)
Comment (zpcore) [reply torch/distributed/tensor/_decompositions.py]: IIUC, basically we are trying to list all possible placements here and then filter them out through decomposition. Drop the combination if there is redistribution needed during the decompose propagation. 

However, if there is no Shard in the original args placements, then we will miss the opportunity to test input placement with Shard. Is this the case?
